### PR TITLE
Port of base_bank based on PCatinean work

### PR DESCRIPTION
**Pull request by nbessi**

### CODE REVIEW

File: l10n_ch_base_bank/bank.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        help='Internal reference'
+    )
+    clearing = fields.Char(
+        string='Clearing number',
+        size=64,
+        help='Swiss unique bank identifier also used in IBAN number'
+    )
+    city = fields.Char(
+        string='City',
+        size=128,
```
Comment (guewen) [inline l10n_ch_base_bank/bank.py]: No reason to put a size on the char fields, excepted if there is a length constraint (e.g. ean13). I don't think there is a Swiss city with more than 128 chars but still the size is useless.

File: l10n_ch_base_bank/bank.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+    ccp = fields.Char(
+        string='CCP',
+        size=64,
+        help="ccp of the bank"
+    )
+
+    @api.constrains('acc_number', 'bank')
+    def _check_ccp_duplication(self):
+        """Ensure validity of input"""
+        p_acc_obj = self.env['res.partner.bank']
```
Comment (guewen) [inline l10n_ch_base_bank/bank.py]: p_acc_obj is not so well named for a `res.partner.bank`, reading the code I thought it was an `account.account`.

File: l10n_ch_base_bank/bank.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+        for bank in self:
+            p_acc_ids = p_acc_obj.search([('bank', '=', bank.id)])
```
Comment (guewen) [inline l10n_ch_base_bank/bank.py]: `p_acc_ids` contains a recordset so you can remove `_ids`

File: l10n_ch_base_bank/bank.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+            if p_acc_ids:
+                check = p_acc_obj._check_ccp_duplication(p_acc_ids)
```
Comment (guewen) [inline l10n_ch_base_bank/bank.py]: `p_acc_ids._check_ccp_duplication()` ?

File: l10n_ch_base_bank/bank.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+                    self._check_5_pos_postal_num(bank.ccp)):
+                return Warning(
+                    _('Please enter a correct postal number. '
+                      '(01-23456-1 or 12345)')
+                )
+        return True
+
+    @api.multi
+    def name_get(self):
+        """Format displaed name"""
```
Comment (guewen) [inline l10n_ch_base_bank/bank.py]: s/displaed/displayed/

File: l10n_ch_base_bank/bank.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+            if not (self._check_9_pos_postal_num(bank.ccp) or
+                    self._check_5_pos_postal_num(bank.ccp)):
+                return Warning(
+                    _('Please enter a correct postal number. '
+                      '(01-23456-1 or 12345)')
+                )
+        return True
+
+    @api.multi
+    def name_get(self):
```
Comment (guewen) [inline l10n_ch_base_bank/bank.py:125]: It seems to me that we should use a computed field `display_name` now, but that's maybe not mandatory.

File: l10n_ch_base_bank/bank.py [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+                        ids += tmp_ids.ids
+                        break
+        # we sort by occurence
+        to_ret_ids = list(set(ids))
+        to_ret_ids = sorted(
+            to_ret_ids,
+            key=lambda x: ids.count(x),
+            reverse=True
+        )
+        self._ids = to_ret_ids
```
Comment (guewen) [inline l10n_ch_base_bank/bank.py]: Are you sure you can do that? I thought you'd rather have to create a new recordset with `self.browse(to_ret_ids)`.

File: l10n_ch_base_bank/invoice.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        """
+        Function to validate a bvr reference like :
+        0100054150009>132000000000000000000000014+ 1300132412>
+        The validation is based on l10n_ch
+        """
+        for invoice in self:
+            if invoice.reference_type == 'bvr' and invoice.state != 'draft':
+                if not invoice.reference:
+                    raise Warning(
+                        _('Error:'),
```
Comment (guewen) [inline l10n_ch_base_bank/invoice.py]: Line to remove

File: l10n_ch_base_bank/bank.py (hunk 6 above)
Comment (nbessi) [inline l10n_ch_base_bank/bank.py:125]: That is a mistery even @xmo-odo did not want to add this in documentation

**Comment by nbessi** [comment]
Thanks for the reviews fixed applied

**Comment by guewen** [comment]
:+1: thanks

**Comment by nbessi** [comment]
Test are missing they can be added in a separtated PR this coed is mandatory for almost all addons and
will slow the porting process

**Comment by nbessi** [comment]
Open issue  https://github.com/OCA/l10n-switzerland/issues/60 about this

File: l10n_ch_base_bank/bank.py [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+        checksum = nums[2]
+        expected_checksum = mod10r(prefix + num)[-1]
+        return expected_checksum == checksum
+
+    def _check_5_pos_postal_num(self, number):
+        """
+        Predicate that checks if a postal number
+        is in format xx-xxxxxx-x is correct,
+        return true if it matches the pattern
+        and if check sum mod10 is ok
```
Comment (yvaucher) [inline l10n_ch_base_bank/bank.py:56]: Wrong doc string. Here we check a 5 pos postal num not a 9 one

xx-xxxxxx-x should be xxxxxx ?

File: l10n_ch_base_bank/bank.py [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+        help='Swiss unique bank identifier also used in IBAN number'
+    )
+    city = fields.Char(
+        string='City',
+        help="City of the bank"
+    )
+    ccp = fields.Char(
+        string='CCP',
+        size=64,
+        help="ccp of the bank"
```
Comment (yvaucher) [inline l10n_ch_base_bank/bank.py]: CCP should be capital letters in help

File: l10n_ch_base_bank/bank.py [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+    and state controls
+
+    """
+    _inherit = 'res.partner.bank'
+    _compile_check_bvr_add_num = re.compile('[0-9]*$')
+
+    bvr_adherent_num = fields.Char(
+        string='Bank BVR adherent number', size=11,
+        help="Your Bank adherent number to be printed "
+             "in references of your BVR."
```
Comment (yvaucher) [inline l10n_ch_base_bank/bank.py]: Missing a space after `...BVR._"` or it will be appened with next line.

File: l10n_ch_base_bank/bank.py [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+                self._check_5_pos_postal_num(p_bank.acc_number) or
+                self._check_9_pos_postal_num(p_bank.acc_number)
+            )
+            bank_check = (
+                self._check_5_pos_postal_num(p_bank.bank.ccp) or
+                self._check_9_pos_postal_num(p_bank.bank.ccp)
+            )
+            if part_bank_check and bank_check:
+                raise exceptions.ValidationError(
+                    _('You can not enter a ccp both on '
```
Comment (yvaucher) [inline l10n_ch_base_bank/bank.py]: CCP in capital letters

File: l10n_ch_base_bank/invoice.py [hunk 13]
```diff
@@ -17,85 +17,77 @@
...
     _inherit = "account.invoice"
 
-    def onchange_partner_id(self, cursor, uid, ids, invoice_type, partner_id,
-                            date_invoice=False, payment_term=False,
-                            partner_bank_id=False, company_id=False):
+    @api.multi
+    def onchange_partner_id(self, invoice_type, partner_id, date_invoice=False,
+                            payment_term=False, partner_bank_id=False,
+                            company_id=False):
         """ Function that is call when the partner of the invoice is changed
```
Comment (yvaucher) [inline l10n_ch_base_bank/invoice.py]: s/call/called

File: l10n_ch_base_bank/invoice.py [hunk 14]
```diff
@@ -17,85 +17,77 @@
...
+        for invoice in self:
             if invoice.type in 'in_invoice':
-                if (invoice.partner_bank_id and
-                        invoice.partner_bank_id.state == 'bvr' and
+                if (invoice.partner_bank_id.state == 'bvr' and
                         invoice.reference_type != 'bvr'):
-                    return False
-        return True
+                    raise exceptions.ValidationError(
+                        _('Invalid Bvr Number (wrong checksum).')
```
Comment (yvaucher) [inline l10n_ch_base_bank/invoice.py]: Bvr -> BVR

File: l10n_ch_base_bank/invoice.py [hunk 15]
```diff
@@ -17,85 +17,77 @@
...
         The validation is based on l10n_ch
         """
-        invoices = self.browse(cr, uid, ids)
-        for invoice in invoices:
+        for invoice in self:
             if invoice.reference_type == 'bvr' and invoice.state != 'draft':
                 if not invoice.reference:
-                    return False
+                    raise exceptions.ValidationError(
+                        _('Invalid Bvr Number (wrong checksum).')
```
Comment (yvaucher) [inline l10n_ch_base_bank/invoice.py]: Bvr -> BVR

Comment (yvaucher) [inline l10n_ch_base_bank/invoice.py]: Shouldn't this message be 'Error: BVR reference is required.' ?

Both checks _check_bvr and _check_reference_type returns the same message. Something is wrong.

File: l10n_ch_base_bank/invoice.py (hunk 14 above)
Comment (yvaucher) [inline l10n_ch_base_bank/invoice.py]: BTW I don't see a checksum check here. There is an error in error message

**Comment by yvaucher** [comment]
See my inline comments.

As a side note, I wonder after our discussion if names as BVR and CCP shouldn't be BVR/ESR and CCP/CP-Konto in english version.

**Comment by nbessi** [comment]
Hello, 

I have done the fixes, One thumb up form the merge

**Comment by yvaucher** [comment]
Seems really better for messages thanks. You will need to update pot file as well. :+1:

**Comment by nbessi** [comment]
Merging Work of @PCatinean, Thanks :+1:

**Pull request merged**